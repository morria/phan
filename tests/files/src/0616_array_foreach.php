<?php

declare(strict_types=1);

/**
 * @param array<int,array<int,array<int,string>>> $array
 */
function test_nested_array_foreach_noshape(array $array) {
    foreach ($array as $key => list(list($_1, $_2), list($_3, $_4), $_5)) {
        echo "{$key}: {$_1}, {$_2}, {$_3}, {$_4}", PHP_EOL;
        // Should infer string for all of these:
        echo intdiv($_1, $_2);
        echo intdiv($_3, $_4);
        echo strlen($_5);  // array<int,string>
    }
}
/**
 * @param array<int,array<int,array<int,string>>> $array
 */
function test_nested_array_foreach_complex(array $array) {
    $newArray = [];
    foreach ($array as $key => $newArray[]) {
        echo strlen($key);  // should infer array
        echo strlen($newArray);  // should infer array<int,array<int,string>> - TODO: Verify array iteration in PHP interpreter
    }
}
