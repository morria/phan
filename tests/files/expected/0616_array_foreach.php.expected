%s:12 PhanTypeMismatchArgumentInternal Argument 1 (numerator) is string but \intdiv() takes int
%s:12 PhanTypeMismatchArgumentInternal Argument 2 (divisor) is string but \intdiv() takes int
%s:13 PhanTypeMismatchArgumentInternal Argument 1 (numerator) is string but \intdiv() takes int
%s:13 PhanTypeMismatchArgumentInternal Argument 2 (divisor) is string but \intdiv() takes int
%s:14 PhanTypeMismatchArgumentInternal Argument 1 (string) is array<int,string> but \strlen() takes string
%s:23 PhanTypeMismatchArgumentInternal Argument 1 (string) is int but \strlen() takes string
%s:24 PhanTypeMismatchArgumentInternal Argument 1 (string) is array<int,array<int,array<int,string>>>|array{} but \strlen() takes string
